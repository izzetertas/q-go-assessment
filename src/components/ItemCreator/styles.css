.itemCreator {
  margin-top: 20px;
}

.itemCreator-input {
  height: 30px;
  width: 300px;
  padding-left: 10px;
  font-size: 16px;
}

.itemCreator-button {
  height: 30px;
  width: 100px;
  margin-left: 10px;
}
