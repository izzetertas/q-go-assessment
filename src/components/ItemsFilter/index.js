import React from 'react';
import PropTypes from 'prop-types';

import FilterLink from './FilterLink';


const ItemsFilter = ({ filterBy, onFilter, totalItems }) => {
  const handleOnClick = (filter) => {
    onFilter(filter);
  };

  return (
    <div>
      <div className="itemsFilter">
        <span className="itemsFilter-total">
          {totalItems > 0 && `${totalItems} items`}
        </span>
        {filterBy !== 'None' &&
          <div>
            <FilterLink
                filterBy={filterBy}
                title="All"
                onClick={handleOnClick} /> | {' '}
            <FilterLink
                filterBy={filterBy}
                title="Done"
                onClick={handleOnClick} /> | {' '}
            <FilterLink
                filterBy={filterBy}
                title="Waiting"
                onClick={handleOnClick} />
          </div>
        }
      </div>

    </div>
  );
};

ItemsFilter.propTypes = {
  filterBy: PropTypes.oneOf([
    'None', 
    'All', 
    'Done', 
    'Waiting'
  ]).isRequired,
  onFilter: PropTypes.func.isRequired
};
ItemsFilter.defaultProps = {
  filterBy: 'None'
};

export default ItemsFilter;
