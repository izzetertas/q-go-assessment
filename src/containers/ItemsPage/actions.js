import {  
  ADD_ITEM,
  DELETE_ITEM,
  TOGGLE_ITEM,
  SET_FILTERBY
} from './constants';

export function addItem(content) {
  return { 
    type: ADD_ITEM, 
    content 
  };
};

export function deleteItem(id){
  return { 
    type: DELETE_ITEM,
    id 
  };
};

export function toggleItem(id) {
  return { 
    type: TOGGLE_ITEM, 
    id 
  };
};

export function setFilterBy(filterBy){ 
  return { 
    type: SET_FILTERBY, 
    filterBy 
  };
};
