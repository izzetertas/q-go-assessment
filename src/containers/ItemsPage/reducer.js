// import { fromJS } from 'immutable';

import {
  ADD_ITEM,
  DELETE_ITEM,
  TOGGLE_ITEM,
  SET_FILTERBY
} from './constants';

export const initialState = { // fromJS({
  items: [
    { id: 1, content: 'Call mum', isDone: false },
    { id: 2, content: 'Buy cat food', isDone: false },
    { id: 3, content: 'Water the plants', isDone: false },
  ],
  filterBy: 'All'
}; // );

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case ADD_ITEM: {
      const nextId = state.items.reduce((id, item) => Math.max(item.id, id), 0) + 1,
        newItem = {
          id: nextId,
          content: action.content,
          isDone: false
        },
        filterBy = state.items.length === 0 ? 'All' : state.filterBy;

      return {
        ...state,
        items: [...state.items, newItem],
        filterBy
      };
    }
    case DELETE_ITEM: {
      const items = state.items.filter(i => i.id !== action.id),
        filterBy = items.length === 0 ? 'None' : state.filterBy;
      return {
        ...state,
        items,
        filterBy
      };
    }
    case TOGGLE_ITEM: {
      return {
        ...state,
        items: state.items.map(item =>
          (item.id === action.id)
            ? {...item, isDone: !item.isDone}
            : item
        )
      };
    }
    case SET_FILTERBY: {
      return {
        ...state,
        filterBy: action.filterBy
      };
    }
    default:
      return state;
  }
}
