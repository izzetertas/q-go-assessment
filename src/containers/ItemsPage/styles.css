.items {
  outline-offset: 20px;
  border: 1px solid #e8e0e0;
  margin: auto;
  max-width: 500px; 
  margin-top: 10px;
}