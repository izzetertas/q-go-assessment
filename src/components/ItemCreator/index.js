import React from 'react';
import PropTypes from 'prop-types';
import './styles.css';

const ItemCreator = ({ onAdd }) => {
  let inputField;
  const handleOnSubmit = e => {
    e.preventDefault();
    if(inputField.value) {
      onAdd(inputField.value);
      inputField.value = '';
    }
  };

  return (
    <div className="itemCreator">
      <form onSubmit={handleOnSubmit}>
        <input
          ref={input => {
            inputField = input;
          }}
          className="itemCreator-input"
          type="text"
          placeholder="What do you need to do?"
        />
        <input
          className="itemCreator-button"
          type="button"
          value="Add Task"
          onClick={handleOnSubmit}
        />
      </form>
    </div>
  );
};
ItemCreator.propTypes = {
  onAdd: PropTypes.func.isRequired,
};

export default ItemCreator;
