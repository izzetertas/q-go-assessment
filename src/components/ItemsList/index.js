import React from 'react';
import PropTypes from 'prop-types';

import './styles.css';
import ItemDetail from '../ItemDetail'


const ItemsList = ({ items, onToggle, onDelete }) => (
  <div className="itemsList">
    <ul className="itemsList-ul">
      {items && items.length < 1 && 
        <p id="items-missing">Add some tasks above.</p>
      }
      {items && items.map(item => 
        <ItemDetail
          key={item.id} 
          item={item} 
          onDelete={id => onDelete(id)}
          onToggle={id => onToggle(id)} />
      )}
    </ul>
  </div>
);

ItemsList.propTypes = {
  items: PropTypes.array.isRequired,
  onToggle: PropTypes.func.isRequired
};

export default ItemsList;
