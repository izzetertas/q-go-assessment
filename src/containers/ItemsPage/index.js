import React, { Component } from 'react';
import { connect } from 'react-redux';

import { addItem, toggleItem, deleteItem, setFilterBy } from './actions';
import ItemCreator from '../../components/ItemCreator';
import ItemsList from '../../components/ItemsList';
import ItemsFilter from '../../components/ItemsFilter';
import './styles.css';


class ItemsPage extends Component {
  handleOnAdd = title => {
    this.props.onAdd(title);
  };
  handleOnToggle = id => {
    this.props.onToggle(id);
  };
  handleOnDelete = id => {
    this.props.onDelete(id);
  };
  handleOnFilter = filterBy => {
    this.props.onFilter(filterBy);
  }
  getFilteredTodos = todo => {
    return this.props.filterBy === 'All' || 
      (this.props.filterBy === 'Done' && todo.isDone) ||
      (this.props.filterBy === 'Waiting' && !todo.isDone)
      ? true
      : false;
  }
  render() {
    return (
      <div className="items">
        <ItemCreator 
          onAdd={this.handleOnAdd}
        />
        <ItemsList 
          items={this.props.items.filter(this.getFilteredTodos)}  
          onToggle={this.handleOnToggle}
          onDelete={this.handleOnDelete}
        />
        <ItemsFilter 
          filterBy={this.props.filterBy}
          onFilter={this.handleOnFilter}
          totalItems={this.props.items.length}
        />
      </div>
    );
  }
}

const mapStateToProps = state => {
  return { 
    items: state.todos.items,
    filterBy: state.todos.filterBy
  };
};

export const mapDispatchToProps = dispatch => ({
  onAdd: newItem => dispatch(addItem(newItem)),
  onToggle: id => dispatch(toggleItem(id)),
  onDelete: id => dispatch(deleteItem(id)),
  onFilter: filterBy => dispatch(setFilterBy(filterBy))
});

export default connect(
  mapStateToProps, 
  mapDispatchToProps
)(ItemsPage);
