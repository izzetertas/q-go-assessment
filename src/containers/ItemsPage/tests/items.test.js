import React from 'react';
import { mount } from 'enzyme';
import configureStore from 'redux-mock-store'
import { Provider } from 'react-redux';
import TestRenderer from 'react-test-renderer';

import ItemsPage, { mapDispatchToProps } from '../index';
import ItemCreator from '../../../components/ItemCreator';
import ItemsList from '../../../components/ItemsList';
import ItemsFilter from '../../../components/ItemsFilter';

const defaultProps = {
  items: [],
  filterBy: 'None'
};

const middlewares = [];
const mockStore = configureStore(middlewares);
const initialState = {
  todos: {
    items: [],
    filterBy: 'None'
  }
};

describe('<ItemsPage />', () => {
  let renderedItem;
  
  beforeEach(() => {
    renderedItem = mount(
      <Provider store={mockStore(initialState)}>
        <ItemsPage />
      </Provider>
    )
  });

  it('should match its empty snapshot', () => {
    expect(TestRenderer.create(
      <Provider store={mockStore(initialState)}>
        <ItemsPage {...defaultProps} />
      </Provider>
    ).toJSON()).toMatchSnapshot();
  });

  it('should render items correctly', () => {
    expect(renderedItem.hasClass('items')).toBeTruthy();
    expect(renderedItem.find(ItemsPage)).toHaveLength(1);
    expect(renderedItem.find(ItemCreator).length).toBe(1);
    expect(renderedItem.find(ItemsList)).toHaveLength(1);
    expect(renderedItem.find(ItemsFilter)).toHaveLength(1);
  });

  it('should call onAdd action with correct result', () => {
    const title = 'add Item title';
    const dispatch = jest.fn();
    mapDispatchToProps(dispatch).onAdd(title);
    expect(dispatch.mock.calls[0][0]).toEqual({ 
      type: 'qgo/assessment/ADD_ITEM', 
      content: title
    });  
  });

  it('should call deleteItem action with correct id', () => {
    const id = 10;
    const dispatch = jest.fn();
    mapDispatchToProps(dispatch).onDelete(id);
    expect(dispatch.mock.calls[0][0]).toEqual({ 
      type: 'qgo/assessment/DELETE_ITEM', 
      id
    });  
  });

  it('should call onToggle action with correct id', () => {
    const id = 10;
    const dispatch = jest.fn();
    mapDispatchToProps(dispatch).onToggle(id);
    expect(dispatch.mock.calls[0][0]).toEqual({ 
      type: 'qgo/assessment/TOGGLE_ITEM', 
      id
    });  
  });

  it('should call setFilterBy action with correct filter', () => {
    const filterBy = 'Done';
    const dispatch = jest.fn();
    mapDispatchToProps(dispatch).onFilter(filterBy);
    expect(dispatch.mock.calls[0][0]).toEqual({ 
      type: 'qgo/assessment/SET_FILTERBY', 
      filterBy
    });  
  });
});
