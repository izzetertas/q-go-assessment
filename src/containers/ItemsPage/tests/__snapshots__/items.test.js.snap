// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ItemsPage /> should match its empty snapshot 1`] = `
<div
  className="items"
>
  <div
    className="itemCreator"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        className="itemCreator-input"
        placeholder="What do you need to do?"
        type="text"
      />
      <input
        className="itemCreator-button"
        onClick={[Function]}
        type="button"
        value="Add Task"
      />
    </form>
  </div>
  <div
    className="itemsList"
  >
    <ul
      className="itemsList-ul"
    >
      <p
        id="items-missing"
      >
        Add some tasks above.
      </p>
    </ul>
  </div>
  <div>
    <div
      className="itemsFilter"
    >
      <span
        className="itemsFilter-total"
      />
    </div>
  </div>
</div>
`;
